consul/resource_consul_configuration_entry.go

```diff
@@ -0,0 +1,176 @@
+package consul
+
+import (
+	"fmt"
+	"strings"
+
+	consulapi "github.com/hashicorp/consul/api"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+)
+
+const (
+	consulConfigEntryServiceDefaults = "service-defaults"
+	consulConfigEntryProxyDefaults   = "proxy-defaults"
+)
+
+func resourceConsulConfigurationEntry() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceConsulConfigurationEntryCreate,
+		Update: resourceConsulConfigurationEntryCreate,
+		Read:   resourceConsulConfigurationEntryRead,
+		Delete: resourceConsulConfigurationEntryDelete,
+
+		Schema: map[string]*schema.Schema{
+			"kind": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice(
+					[]string{
+						consulConfigEntryServiceDefaults,
+						consulConfigEntryProxyDefaults,
+					},
+					false,
+				),
+			},
+
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"protocol": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"config": {
+				Type:     schema.TypeMap,
+				Optional: true,
+			},
+
+			"token": {
+				Type:      schema.TypeString,
+				Optional:  true,
+				Sensitive: true,
+			},
```

Comment:
Hmm I see why mixing all the fields in every config Entry type we add gets a bit messy here. This is fine for now but I expect the number of fields here to increase a lot as we add more options to Connect etc.

Is there a way we can make this more generic so it doesn't need constant maintenance to keep up with the config syntax for all these types?

For example in Consul core config we allow bootstrapping these from HCL but we internal just type it as `map[string]interface{}` so we don't have to maintain the config entry schema in lots of different places.

I'm not super familiar with TF providers though so I'm not sure what is possible and what the downsides are of making it more generic. What do you think @remilapeyre ?

Reply:
For example, the beta release we will cut this week adds 3 more "kinds" here that have another 10+ top level fields we'd have to add to this schema my hope was we could keep this generic and then have it just work for new types automatically.

Reply:
Is it possible for the top-level Schema just to be a `schema.TypeMap`? Or to define kind and name but then have TF just present all the other fields generically?